fili-core/src/main/java/com/yahoo/bard/webservice/web/apirequest/DataApiRequestValueObject.java

```diff
@@ -0,0 +1,772 @@
+// Copyright 2019 Oath Inc.
+// Licensed under the terms of the Apache license. Please see LICENSE.md file distributed with this work for terms.
+package com.yahoo.bard.webservice.web.apirequest;
+
+import com.yahoo.bard.webservice.data.dimension.Dimension;
+import com.yahoo.bard.webservice.data.dimension.DimensionDictionary;
+import com.yahoo.bard.webservice.data.dimension.DimensionField;
+import com.yahoo.bard.webservice.data.metric.LogicalMetric;
+import com.yahoo.bard.webservice.data.time.Granularity;
+import com.yahoo.bard.webservice.druid.model.builders.DruidFilterBuilder;
+import com.yahoo.bard.webservice.druid.model.filter.Filter;
+import com.yahoo.bard.webservice.druid.model.having.Having;
+import com.yahoo.bard.webservice.druid.model.orderby.OrderByColumn;
+import com.yahoo.bard.webservice.table.LogicalTable;
+import com.yahoo.bard.webservice.util.UnmodifiableLinkedHashMap;
+import com.yahoo.bard.webservice.util.UnmodifiableLinkedHashSet;
+import com.yahoo.bard.webservice.web.ApiFilter;
+import com.yahoo.bard.webservice.web.ApiHaving;
+import com.yahoo.bard.webservice.web.ResponseFormatType;
+import com.yahoo.bard.webservice.web.filters.ApiFilters;
+import com.yahoo.bard.webservice.web.filters.UnmodifiableApiFilters;
+import com.yahoo.bard.webservice.web.util.PaginationParameters;
+
+import org.joda.time.DateTimeZone;
+import org.joda.time.Interval;
+
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.LinkedHashMap;
+import java.util.LinkedHashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+import javax.ws.rs.core.Response;
+
+/**
+ * And immutable POJO implementation of {@link DataApiRequest} contract. All data is provided through the constructor
+ * and set, with very minor or no transformations occurring on the data. Unlike the old implementation
+ * ({@link com.yahoo.bard.webservice.web.apirequest.DataApiRequestImpl}), this implementation does not build any of
+ * tis components and all components must be fully built at creation time.
+ */
+public class DataApiRequestValueObject implements DataApiRequest {
+
+    private final LogicalTable table;
+    private final Granularity granularity;
+    private final LinkedHashSet<Dimension> dimensions;
+    private final LinkedHashMap<Dimension, LinkedHashSet<DimensionField>> perDimensionFields;
+    private final LinkedHashSet<LogicalMetric> metrics;
+    private final List<Interval> intervals;
+    private final ApiFilters apiFilters;
+    private final LinkedHashMap<LogicalMetric, Set<ApiHaving>> havings;
+    private final OrderByColumn dateTimeSort;
+    private final LinkedHashSet<OrderByColumn> standardSorts;
+    private final LinkedHashSet<OrderByColumn> allSorts;
```

Comment:
Why are we keeping all three of these?  We should either keep the separate fields or the merged one.